### PR TITLE
Adds support xen domU lightweight containers

### PR DESCRIPTION
**Pull request by angerman**
When runnign nixOS in a xen setup without it's own
kernel (the kernel and initrd are provided by the host)
we need to use init-scripts (/sbin/init) instead of grub
and have no kernel.

###### Motivation for this change

This adds support to run nixOS on DomainFactorys
(df.eu) JiffyBoxs with the "PV-Kernel". It should also
allow to run nixOS on similar setups xen guest setups
where the host provides the kernel.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
btw. we also have a section in our wiki on how to install nixos on different providers: https://nixos.wiki/wiki/NixOS_friendly_hosters

File: nixos/modules/system/boot/loader/init-script/init-script.nix [hunk 1]
```diff
@@ -22,7 +22,10 @@ in
     boot.loader.initScript = {
 
       enable = mkOption {
-        default = false;
+        # When being booted in as a xen domU
+        # with a provided kernel and initrd,
+        # we need the /sbin/init script.
+        default = config.boot.usePvKernel;
```
Comment (Mic92) [inline nixos/modules/system/boot/loader/init-script/init-script.nix]: Every user with PV Kernel would now have to regenerate the manual when you override the default option.
I would rather override this in a module or profile using `mkOverride`.
 We also have `nixos/modules/virtualisation/xen-domU.nix` file, how is this related to your property?

Comment (angerman) [reply nixos/modules/system/boot/loader/init-script/init-script.nix]: I did this here because the same was done for `grub` and `isContainer` in
https://github.com/angerman/nixpkgs/blob/608eaeaeab10593a3f661e9e1012286febbaeb45/nixos/modules/system/boot/loader/grub/grub.nix#L92-L98

The `xen-domU.nix` sets up a xen guest, but expects it to have a boot loader. The situation I am in is in between what xen-domU offers and what the isContainer flags does. However the isContainer flag expects to be a container on a nixOS host.

**Comment by angerman** [comment]
> btw. we also have a section in our wiki on how to install nixos on different providers: https://nixos.wiki/wiki/NixOS_friendly_hosters

Right. domainfactory is not listed. And I'm not sure I would advise to use domainfactory with nixOS right now. With the missing PV Kernel support, and their only active Xen and PV kernel offerings for their JiffyBox's it's more of a challenge to get NixOS working.

I just happen to have a JiffyBox already, otherwise I would likely have chosen digital ocean or similar.

File: nixos/modules/system/boot/loader/init-script/init-script.nix (hunk 1 above)
Comment (Mic92) [reply nixos/modules/system/boot/loader/init-script/init-script.nix]: `isContainer` probably was also not the best design decision, because it clutters every module.

Comment (angerman) [reply nixos/modules/system/boot/loader/init-script/init-script.nix]: I’ll need some guidance here. How do I set initScripts.enabled to true if usePvKernel is set to true? Unless the user configured usePvKernel = true; initScripts.enabled = false;

usePvKernel basically implies that initScripts are used as the kernels initrd will very likely expect to find /sbin/init as the standard common entry point for the guest.

Comment (angerman) [reply nixos/modules/system/boot/loader/init-script/init-script.nix]: The key being that nixOS has no control over the kernel or initrd; and the host has no idea about the system and as such treats it as a generic linux.

Comment (Mic92) [reply nixos/modules/system/boot/loader/init-script/init-script.nix]: I opened that issue here to discuss the problem: https://github.com/NixOS/nixpkgs/issues/33940

Comment (Mic92) [reply nixos/modules/system/boot/loader/init-script/init-script.nix]: We would need for each module where `isContainer` is set a local option to disable this aspect and then set the same option in your pv-kernel module or in a generic container profile. In the linked issue I propose a kernel-less module.

Comment (angerman) [reply nixos/modules/system/boot/loader/init-script/init-script.nix]: That seems to open up another whole can worms. This feels a bit outside of this PR.

**Comment by angerman** [comment]
What's the status of this? As we are tracking the "optimization" in #33940, can we in the meantime accept and merge this?

**Comment by Mic92** [comment]
https://github.com/NixOS/nixpkgs/issues/33940#issuecomment-358387958 looks easy to implement to me.
If I accept this change now it will just increase technical dept and nobody will attempt to fix it.
